### ISSUE TITLE
test_onedim.TestFreeFlame fails on OSX 10.15.6

### ISSUE THREAD

**Issue by santoshshanbhogue** [Cantera/cantera#934]
**Problem description**
test_onedim.TestFreeFlame fails on OSX

**Steps to reproduce**
Run `scons test `

**Behavior**
======================================================================
FAIL: test_write_hdf (cantera.test.test_onedim.TestFreeFlame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/santosh1/packages/canteraSource/build/python/cantera/test/test_onedim.py", line 649, in test_write_hdf
    meta = f.read_hdf(filename)
  File "/Users/santosh1/packages/canteraSource/build/python/cantera/onedim.py", line 635, in read_hdf
    self.settings = meta
  File "/Users/santosh1/packages/canteraSource/build/python/cantera/onedim.py", line 663, in settings
    setattr(self, key, s[key])
  File "interfaces/cython/cantera/onedim.pyx", line 1411, in cantera._cantera.Sim1D.fixed_temperature.__set__
cantera._cantera.CanteraError: 
***********************************************************************
CanteraError thrown by StFlow::setupGrid:
grid points must be monotonically increasing
***********************************************************************

**System information**

- Cantera version: [2.5.0b1 commit # 2cd06194e]
- OS: OSX 10.15.6 Catalina
- Python/MATLAB/other software versions: XCode 11.7, python 3.7, numpy 1.17

**Attachments**

**Additional context**
@bryanwweber could reproduce these, but the CI could not

**Comment by ischoegl** [comment]
@bryanwweber ... is there a way for CI to be tweaked so this can be reproduced? The issue seems to be only occurring for OSX.

@santoshshanbhogue ... per your user group post, I believe there was a second failure also?

Having written the unit test for the other failure ... this checks the ability to recreate a `StFlow` object from data saved to HDF from the Python front end. The error points at a grid point being added where it shouldn’t be (I am basing this on one of  @bryanwweber ’s observations In the UG post) ... this could be caused by the code not being able to identify an already existing point as the one that should be used based on numerical criteria. I have some ideas of what may be awry but would rather not speculate beyond what I wrote above (I cannot reproduce on my tool chains).

**Comment by santoshshanbhogue** [comment]
@ischoegl Here is the log of the second failure:

======================================================================
FAIL: test_ion_profile (cantera.test.test_onedim.TestIonBurnerFlame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/santosh1/packages/canteraSource/build/python/cantera/test/test_onedim.py", line 1253, in test_ion_profile
    self.assertNear(max(self.sim.E), 552.33, 1e-2)
  File "/Users/santosh1/packages/canteraSource/build/python/cantera/test/utilities.py", line 57, in assertNear
    self.fail(message)
AssertionError: AssertNear: 580.09163013629 - 552.33 = 27.761630136288
Relative error of 4.903055e-02 exceeds rtol = 1.000000e-02

----------------------------------------------------------------------

**Comment by bryanwweber** [comment]
@ischoegl I don't know what to change about the CI to reproduce the failure. As best I can tell, my build toolchain is the same as CI... The grid point being added is a zero as the last element of the array, due to the right hand boundary being processed. I'm not confident about how the domains work to know whether that point should be there or not.

**Comment by ischoegl** [comment]
@bryanwweber ... that is *very* odd indeed. I was hoping that there’d be some difference in CI that would explain this issue. Without seeing more about the problem it’s difficult to untangle this. Also, any ideas about the other failure? I’m not sure what the normal discrepancy is there, but 5% relative error appears large.

@santoshshanbhogue ... were there any other customizations when compiling? I.e. could you post your `cantera.conf`?

**Comment by ischoegl** [comment]
> The grid point being added is a zero as the last element of the array, due to the right hand boundary being processed. 

I'd need to see more output to follow (what are the values of `key` and `s[key]` in `setattr(self, key, s[key])`)? To me, the error stack points towards setting the fixed point of the free flame in `cantera._cantera.Sim1D.fixed_temperature.set`, i.e.

https://github.com/Cantera/cantera/blob/2cd06194e80aa0f43cba36575ecf1cdc6c073391/interfaces/cython/cantera/onedim.pyx#L1403-L1411

where a potential culprit is the comparison of two floats on line 449 in `Sim1D.cpp` where an existing point may not be detected (i.e. the comparison uses an exact equality - which is problematic for floats)

https://github.com/Cantera/cantera/blob/2cd06194e80aa0f43cba36575ecf1cdc6c073391/src/oneD/Sim1D.cpp#L449-L468

In other words: `read_hdf` tries to recreate an old solution, which had already created the grid point with the temperature in question: `addnewpt` should be `false` in this case.

I cannot verify this on my end though ...

**Comment by santoshshanbhogue** [comment]
@ischoegl my cantera.conf is at the end of this block. I also did a "sensitivity" test with py38/py37, boost 1.73/1.55, numpy 1.19/1.17 and the error did not go away

@bryanwweber when you reproduced the failures, did you use Catalina 10.15.6 or 10.15.5? Does the CI use 10.15.5? 

prefix = '/Users/santosh1/packages/ct_dev_version'
python_package = 'full'
matlab_toolbox = 'y'
matlab_path = '/Applications/MATLAB_R2019a.app/'
f90_interface = 'n'
system_eigen = 'n'
system_fmt = 'n'
system_sundials = 'n'

**Comment by ischoegl** [comment]
@santoshshanbhogue ... thank you. Outside what I write above I’m stumped. You *could* try to test my theory by replacing the equality check I pointed to with checking a relative tolerance instead? I have a suspicion that this is caused by truncated values, but there are probably other points in the code that may cause this also.

**Comment by bryanwweber** [comment]
I'm on 10.14. I'm not sure what the CI uses, but obviously this is reproduced on at least two versions of macOS.

**Comment by santoshshanbhogue** [comment]
@ischoegl I ran a debugger and I see a couple of issues in this section of the code:  https://github.com/Cantera/cantera/blob/07dac7be8244ab6bea23a48096f891bdea45f3ab/interfaces/cython/cantera/test/test_onedim.py#L639

In this line: https://github.com/Cantera/cantera/blob/07dac7be8244ab6bea23a48096f891bdea45f3ab/interfaces/cython/cantera/test/test_onedim.py#L644

`refine=False` does not make its way through to the 'solve_mix' function. In the latter the `refine_grid` is set to `True` 

Now, here is the output of the debugger with a breakpoint at this line: https://github.com/Cantera/cantera/blob/07dac7be8244ab6bea23a48096f891bdea45f3ab/interfaces/cython/cantera/test/test_onedim.py#L649

`(Pdb) p self.sim.grid
array([0. , 0.4, 0.6, 0.7, 0.8, 1. , 1.2, 1.6, 2. ])
(Pdb) p f.grid
array([0.  , 0.02, 0.04, 0.06, 0.08, 0.1 ])
`

Notice that f and self.sim have different values... Based on this, the code should fail at line 650...

**Comment by ischoegl** [comment]
@santoshshanbhogue ... thanks for the output. On line 649 of the unit test, everything within `f` Is essentially replaced by the contents of the HDF file, so depending on when you query the values, there may be discrepancies. Once you reach line 650, everything should match (which is confirmed by CI). I am afraid that the root cause of the failure is elsewhere, likely deep inside the C++ code ...

**Comment by ischoegl** [comment]
@santoshshanbhogue ... here's something that may be easier to debug. Run the `adiabatic_flame.py` script in `examples/onedim`, which will generate `adiabatic_flame.h5`. Then, run the following script within the folder:
```Python
import cantera as ct

# create an empty FreeFlame object
gas = ct.Solution('h2o2.yaml')
f = ct.FreeFlame(gas)

# load previous solution into FreeFlame
f.read_hdf('adiabatic_flame.h5') # <--- this presumably fails on OSX
```
Running the same on iPython (plus some extra output), this is what I get on Ubuntu 18.04
```
In [1]: import cantera as ct
   ...: gas = ct.Solution('h2o2.yaml')
   ...: f = ct.FreeFlame(gas)
   ...: f.grid.shape
   ...:
Out[1]: (6,)

In [2]: f.read_hdf('adiabatic_flame.h5') # <--- this presumably fails on OSX
Out[2]: 
{'Sim1D_type': 'FreeFlame',
 'cantera_version': '2.5.0b1',
 'curve': 0.12,
 'date': 'Sat, 12 Sep 2020 04:36:17 +0000',
 'description': 'solution with mixture-averaged transport',
 'energy_enabled': True,
 'fixed_temperature': 698.1678485027371,
 'git_commit': '2cd06194e',
 'max_grid_points': 1000,
 'max_time_step_count': 500,
 'prune': 0.0,
 'radiation_enabled': False,
 'ratio': 3.0,
 'slope': 0.06,
 'soret_enabled': False,
 'transport_model': 'Mix'}

In [3]: f.grid.shape
Out[3]: (140,)

In [4]: f.fixed_temperature
Out[4]: 698.1678485027371
```
The error stack more or less points at the `fixed_temperature` entry, which is re-set when HDF data are loaded, and this may be what's causing the issue.

**Comment by santoshshanbhogue** [comment]
Strangely, this runs fine. Here is my output

'{'Sim1D_type': 'FreeFlame',
 'cantera_version': '2.5.0b1',
 'curve': 0.12,
 'date': 'Sat, 12 Sep 2020 00:59:06 -0400',
 'description': 'solution with mixture-averaged transport',
 'energy_enabled': True,
 'fixed_temperature': 698.1678485027371,
 'git_commit': '2cd06194e',
 'max_grid_points': 1000,
 'max_time_step_count': 500,
 'prune': 0.0,
 'radiation_enabled': False,
 'ratio': 3.0,
 'slope': 0.06,
 'soret_enabled': False,
 'transport_model': 'Mix'}'

**Comment by ischoegl** [comment]
Hmm. This is somewhat infuriating ...

**Comment by santoshshanbhogue** [comment]
If it helps, here is the h5 generated by the failing code:
https://www.dropbox.com/s/rhx0gfphukthw9a/onedim-write_hdf.h5?dl=0

I can open the h5 file in MATLAB, can you do this with `read_hdf` on Ubuntu/Windows?

**Comment by ischoegl** [comment]
@santoshshanbhogue ... alright, this actually helps!
```
In [1]: import cantera as ct
   ...: gas = ct.Solution('h2o2.yaml')
   ...: f = ct.FreeFlame(gas)

In [2]: f.read_hdf('onedim-write_hdf.h5')
---------------------------------------------------------------------------
CanteraError                              Traceback (most recent call last)
<ipython-input-4-ebf642b7ffdb> in <module>()
----> 1 f.read_hdf('onedim-write_hdf.h5')

/usr/local/lib/python3.6/dist-packages/cantera/onedim.py in read_hdf(self, filename, group, restore_boundaries)
    633             self.from_solution_array(arr, domain=d)
    634 
--> 635         self.settings = meta
    636 
    637         return meta

/usr/local/lib/python3.6/dist-packages/cantera/onedim.py in settings(self, s)
    661         attr = attr & set(s.keys())
    662         for key in attr:
--> 663             setattr(self, key, s[key])
    664 
    665         # refine criteria

/src/interfaces/cython/cantera/onedim.pyx in cantera._cantera.Sim1D.fixed_temperature.__set__()
   1409             return self.sim.fixedTemperature()
   1410         def __set__(self, T):
-> 1411             self.sim.setFixedTemperature(T)
   1412 
   1413     property fixed_temperature_location:

CanteraError: 
***********************************************************************
CanteraError thrown by StFlow::setupGrid:
grid points must be monotonically increasing
***********************************************************************
```
I.e. this appears to be the same as what you see, correct?

Also, I confirmed that this happens while setting `fixed_temperature`. The complete settings that are restored are:
```
{'Sim1D_type': 'FreeFlame', 'cantera_version': '2.5.0b1', 'curve': 0.2, 'date': 'Sat, 12 Sep 2020 00:50:12 -0400', 'description': 'mixture-averaged simulation', 'energy_enabled': True, 'fixed_temperature': 936.5714010534294, 'git_commit': '2cd06194e', 'max_grid_points': 1000, 'max_time_step_count': 500, 'prune': 0.0, 'radiation_enabled': False, 'ratio': 3.0, 'slope': 0.3, 'soret_enabled': False, 'transport_model': 'Mix'}
```

**Comment by ischoegl** [comment]
@santoshshanbhogue / @bryanwweber ... one of the two “OSX” bugs is in `Sim1D::setFixedTemperature` as expected. For the dataset that is generated by OSX, the second comparison on line 455 should fail (values are supposed to be equal), but the actual difference is `1.13687e-13` ... for a value of `936.571`,  i.e. all of this is because of truncation error.

As mentioned, comparing floats is problematic and can yield different results for different compilers/machines, see [this FAQ](http://www.cs.technion.ac.il/users/yechiel/c++-faq/floating-point-arith.html).

**PS:** will take care of this issue one of these days; the other failure remains to be tracked down (I opened a separate issue). Per UG @decaluwe had been noticing it for a while.

**Comment by ischoegl** [comment]
@santoshshanbhogue / @bryanwweber ... I'm fairly sure the issue is fixed by #936. Feel free to pull from my fork and check, i.e. [`ischoegl:fix-osx-failure-in-hdf`](https://github.com/ischoegl/cantera/tree/fix-osx-failure-in-hdf).

**Comment by santoshshanbhogue** [comment]
Solved for OSX 10.15.16, see https://github.com/Cantera/cantera/pull/936#issuecomment-691701439

**Comment by ischoegl** [comment]
Thanks for checking! I am not planning to work on the other issues (cannot reproduce locally and it doesn’t occur in code I have worked on).
